packages/builder/src/lib/script-checkpoints.ts

```diff
@@ -1,105 +1,310 @@
+  AccountDef,
+  AlgobDeployer,
+  AlgobRuntimeEnv,
+  ASAInfo,
+  ASCInfo,
+  CheckpointData,
+  ScriptCheckpoints,
+  ScriptNetCheckpoint
+} from "../types";
 
 export const scriptsDirectory = "scripts";
 const artifactsPath = "artifacts";
+const checkpointFileSuffix = ".cp.yaml";
 
 export function toCheckpointFileName (scriptName: string): string {
-  return path.join(artifactsPath, scriptName + ".cp.yaml");
+  return path.join(artifactsPath, scriptName + checkpointFileSuffix);
 }
 
-export function appendToCheckpoint (
-  prev: ScriptCheckpoint,
-  networkName: string,
-  append: ScriptNetCheckpoint
-): ScriptCheckpoint {
-  if (prev[networkName]) {
-    prev[networkName].timestamp = append.timestamp;
-    prev[networkName].metadata = Object.assign({}, prev[networkName].metadata, append.metadata);
-    return prev;
-  }
-  prev[networkName] = append;
-  return prev;
+export function registerASA (
+  cp: ScriptNetCheckpoint, name: string, creator: string): ScriptNetCheckpoint {
+  cp.asa[name] = { creator: creator };
+  return cp;
+}
+
+export function registerASC (
+  cp: ScriptNetCheckpoint, name: string, creator: string): ScriptNetCheckpoint {
+  cp.asc[name] = { creator: creator };
+  return cp;
 }
 
-export function mergeCheckpoints (
-  prev: ScriptCheckpoint,
-  curr: ScriptCheckpoint
-): ScriptCheckpoint {
-  const keys: string[] = Object.keys(curr);
-  return keys.reduce((out: ScriptCheckpoint, key: string) => {
-    return appendToCheckpoint(out, key, curr[key]);
-  }, prev);
+export class ScriptNetCheckpointImpl implements ScriptNetCheckpoint {
+  timestamp: number;
+  metadata: { [key: string]: string };
+  asa: { [name: string]: ASAInfo };
+  asc: { [name: string]: ASCInfo };
+
+  constructor (metadata?: {[key: string]: string}) {
+    this.timestamp = +new Date();
+    this.metadata = (metadata === undefined ? {} : metadata);
+    this.asa = {};
+    this.asc = {};
+  }
 }
 
-export function appendEnv (
-  prev: ScriptCheckpoint,
-  runtimeEnv: AlgobRuntimeEnv
-): ScriptCheckpoint {
-  return appendToCheckpoint(prev, runtimeEnv.network.name, createNetCheckpoint());
+export function appendToCheckpoint (
+  checkpoints: ScriptCheckpoints, networkName: string, append: ScriptNetCheckpoint): ScriptCheckpoints {
+  const orig = checkpoints[networkName];
+  if (!orig) {
+    checkpoints[networkName] = Object.assign({}, append);
+    return checkpoints;
+  }
+  orig.timestamp = append.timestamp;
+  orig.metadata = Object.assign(
+    {}, orig.metadata, append.metadata
+  );
+  const allAssetNames = Object.keys(append.asa).concat(Object.keys(append.asc));
+  for (const assetName of allAssetNames) {
+    if ((orig.asa[assetName] && !deepEqual(orig.asa[assetName], append.asa[assetName])) ||
+      (orig.asc[assetName] && !deepEqual(orig.asc[assetName], append.asc[assetName]))) {
+      throw new BuilderError(
+        ERRORS.BUILTIN_TASKS.CHECKPOINT_ERROR_DUPLICATE_ASSET_DEFINITION,
+        { assetName: assetName });
+    }
+  }
```

Comment:
The asset duplicate condition should be only with name -- we shouldn't do `deepEqual` check (we can remove that dependency). Only name must be unique.